### PR TITLE
A 879/kademlia testapp

### PR DESCRIPTION
**Pull request by jtownson**
This is a test app for kademlia, creating a simple kademlia console which allows the user to manually perform kademlia queries as well as display and modify the state of a kademlia node.

To get started using the app, see the readme.md in the examples directory.

### CODE REVIEW

File: scalanet/examples/src/io/iohk/scalanet/kconsole/PureConfigReadersAndWriters.scala [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+import io.iohk.scalanet.peergroup.kademlia.KRouter.NodeRecord
+import pureconfig.{ConfigReader, ConfigWriter}
+import pureconfig.ConvertHelpers.catchReadError
+import pureconfig.configurable.{genericMapReader, genericMapWriter}
+import pureconfig.generic.auto._
+import scodec.bits.BitVector
+
+object PureConfigReadersAndWriters {
+  implicit val bitvectorReader: ConfigReader[BitVector] =
+    ConfigReader[String].map(BitVector.fromValidHex(_))
```
Comment (krcz) [inline scalanet/examples/src/io/iohk/scalanet/kconsole/PureConfigReadersAndWriters.scala:15]: Don't you want to apply `catchReadError` here as well?

File: scalanet/examples/src/io/iohk/scalanet/kconsole/CommandParser.scala [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+  sealed trait Command {
+    def applyTo(kRouter: KRouter[InetMultiAddress]): String
+  }
+
+  object Command {
+    import scala.concurrent.ExecutionContext.Implicits.global
+
+    case class GetCommand(nodeId: BitVector) extends Command {
+      override def applyTo(kRouter: KRouter[InetMultiAddress]): String = {
+        val p = Promise[String]()
```
Comment (krcz) [inline scalanet/examples/src/io/iohk/scalanet/kconsole/CommandParser.scala:24]: Nitpicking: why use `Promise` when 'kRouter.get(nodeId).transform` should do the work?

**Review by krcz — APPROVED** [review]
One additional comment: when initializing with wrong bootstraping ID, node entry is duplicated:

```
> dump
{"id":"64bc4ee0fa5c8d739150245e630f1a2847d5ffc4","messaging-address":"localhost:37551","routing-address":"localhost:33481"}
{"id":"1624da0bc90d1eab099218bfba6f273a70285848","messaging-address":"localhost:36349","routing-address":"localhost:34673"}
{"id":"a98e6fa629b7b4ae679748ad65915f1bf1178ac0","messaging-address":"localhost:36349","routing-address":"localhost:34673"}
```

**Comment by jtownson** [comment]
Merged via command line `git merge --squash`

**Pull request closed without merging**